internal/keepers/service.go

```diff
@@ -132,25 +146,101 @@ func (s *onDemandUpkeepService) CheckUpkeep(ctx context.Context, keys ...types.U
 func (s *onDemandUpkeepService) start() {
 	// TODO: if this process panics, restart it
 	go s.cacheCleaner.Run(s.cache)
+	go func() {
+		if err := s.runSamplingUpkeeps(); err != nil {
+			s.logger.Fatal(err)
+		}
+	}()
 }
 
 func (s *onDemandUpkeepService) stop() {
 	close(s.stopProcs)
 	close(s.cacheCleaner.stop)
 }
 
-func (s *onDemandUpkeepService) parallelCheck(ctx context.Context, keys []types.UpkeepKey, filters ...func(types.UpkeepKey) bool) ([]*types.UpkeepResult, error) {
-	samples := newSyncedArray[*types.UpkeepResult]()
+func (s *onDemandUpkeepService) runSamplingUpkeeps() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	heads := make(chan types.BlockKey, 1)
+	defer close(heads)
+
+	// Start the sampling upkeep process for heads
+	go func() {
+		for range heads {
+			s.handleIncomingHead(ctx)
+		}
+	}()
+
+	// Cancel context when receiving the stop signal
+	go func() {
+		<-s.stopProcs
+		cancel()
+	}()
+
+	return s.headSubscriber.OnNewHead(ctx, func(blockKey types.BlockKey) {
+		// This is needed in order to do not block the process when a new head comes in.
+		// The running upkeep sampling process should be finished first before starting
+		// sampling for the next head.
+		select {
+		case heads <- blockKey:
+		default:
+		}
+	})
+}
+
+// handleIncomingHead performs checking upkeep logic for all eligible keys of the given head
+func (s *onDemandUpkeepService) handleIncomingHead(ctx context.Context) {
```

Comment:
nit. can we rename this function to indicate it doesn't really use the incoming head, instead always uses the latest? 

maybe `processLatestHead` of something?